perf(interpreter): use an more effective memcpy for memory load and store in interpreter

wenym1 opened the pull request:
Currently, in interpreter mode, when handling op code related to memory load and store, we use a simple `memcpy` to load and store native memory. However, the length of such memory copy in `WasmEdge` is mostly (even all) the power of 2, like 1, 2, 4, 8 and etc, and when we copy memory of this size, it is more effective to load and store the memory as a whole, which is as a `uint8`, `uint16`, `uint32`, `uint64` and etc. The default memcpy simply read and store the memory byte by byte. In this PR, such improvement is implemented in a function `effective_memcpy`.

Besides, the `Length` of memory to be operated on can also be resolved statically at compile time. Therefore, I move the `Length` and `BitWidth` argument to the template argument so that the compiler can better optimized our code and reduce the number of branch.

The optimization in this PR is mentioned in #1293.

With this PR, the total cost to run a rust python benchmark reduces from 119 seconds to 111 seconds on MacOS M1 chip.

wenym1 commented:
@hydai may also want to review the code.

q82419 commented on include/runtime/instance/memory.h:
```diff
@@ -241,16 +270,11 @@ class MemoryInstance {
-          ErrInfo::InfoBoundary(Offset, Length, Offset + sizeof(T) - 1));
-      return Unexpect(ErrCode::MemoryOutOfBounds);
-    }
+    static_assert(Length <= sizeof(T));
```
Please don't remove the boundary checking here.
For the memory load out of boundary here, we should return error code but not assert.

wenym1 replied:
Both the `Length` and `sizeof(T)` can be known at compile time and `static_assert` will do the check at compile time and fail the compile if the condition is not met.

Do we have any cases and any op that will explicitly load and store out of bound memory in our WasmEdge runtime code? If not, it will be beneficial to use `static_assert` here. First it can prevent us from writing incorrect code by accident. Second it avoids an unnecessary comparison at runtime and avoids a potential branch miss.

q82419 replied:
Although the `Length` can be known at compile time, the instructions can only be known at runtime.
For WASM, we can only get the `Length` by different instructions at runtime and output the error message.

q82419 replied:
Update:

Sorry that I misunderstood the code.
The boundary checking will be done in the following 2 lines.
You're right, this can be checked with `static_assert`. Thanks.

q82419 commented on include/runtime/instance/memory.h:
```diff
@@ -294,16 +320,11 @@ class MemoryInstance {
-          ErrInfo::InfoBoundary(Offset, Length, Offset + sizeof(T) - 1));
-      return Unexpect(ErrCode::MemoryOutOfBounds);
-    }
+    static_assert(Length <= sizeof(T));
```
As above.

wenym1 replied:
Same reason to use `static_assert` as above.

q82419 replied:
As explained above.

q82419 commented on include/runtime/instance/memory.h:
```diff
@@ -261,16 +285,18 @@ class MemoryInstance {
       } else {
         if constexpr (sizeof(T) > 8) {
-          assuming(sizeof(T) == 16);
+          static_assert(sizeof(T) == 16);
```
Please use `assuming` here instead of `static_assert`. This will fit for both debug and release versions.

wenym1 replied:
Same reason to use static_assert as above.

wenym1 replied:
I have resumed to use the original `assume` instead. 

Thanks for pointing it out. I didn't know the if constexpr restriction.

ibmibmibm commented on include/runtime/instance/memory.h:
```diff
@@ -261,16 +285,18 @@ class MemoryInstance {
       if constexpr (std::is_floating_point_v<T>) {
         // Floating case. Do the memory copy.
-        std::memcpy(&Value, &DataPtr[Offset], sizeof(T));
+        *(reinterpret_cast<T *>(&Value)) =
```
Use `std::memcpy` instead.

ibmibmibm commented on include/runtime/instance/memory.h:
```diff
@@ -261,16 +285,18 @@ class MemoryInstance {
           Value = 0;
-          std::memcpy(&Value, &DataPtr[Offset], Length);
+
+          effective_memcpy<Length>(&Value, &DataPtr[Offset]);
```
Use `std::memcpy` instead.

ibmibmibm commented on include/runtime/instance/memory.h:
```diff
@@ -294,16 +320,11 @@ class MemoryInstance {
   ///
   /// \returns void when success, ErrCode when failed.
-  template <typename T>
+  template <typename T, uint32_t const Length = sizeof(T)>
```
`const` in a template parameter definition is meanless.

wenym1 replied:
I have removed the `const`

wenym1 commented on include/runtime/instance/memory.h:
```diff
@@ -31,6 +31,35 @@ namespace WasmEdge {
 namespace Instance {
 
+template <uint32_t const Length>
+void effective_memcpy(void *dest, const void *src) {
```
If we use `volatile` keyword for the `reinterpret_cast`, can we possibly ignore the alignment check? In the CI, all unit tests succeed in debug mode but some fail with seg fault in release mode. Probably in release mode some optimizations are performed provided the assumption that all memory address are aligned. If we mark the cast as `volatile`, can we fix it?

wenym1 replied:
I have added an alignment check in the `effective_memcpy`. When the `src` and `dest` is not properly aligned, it will use memcpy instead. 

Now the previous failed CI can all get passed. I also ran the rustpython benchmark, and the performance is basically the same as the one without the alignment check, which means adding the alignment check can still gain almost the same performance improvement. Therefore I think we can still use the `effective_memcpy` instead.

wenym1 replied:
In this PR, we only have two main changes, refactor the code with const template and introduce effective_memcpy. And I just ran my benchmark again under different combination on my macbook with M1 chip.
| time | run1 | run2 | run3 | avg |
| --- | --- | --- | --- |  --- |
|Current | 118.08s | 117.88s | 118.87s | 118.28s |
|Current with effective_memcpy| 116.11s | 115.41s | 115.69s | 115.74s|
| const template with std::memcpy | 111.03s | 111.15s | 111.38s | 111.19s |
| const template with effective_memcpy | 110.87s | 111.10s | 112.04s | 111.34s| 

Seems that the performance is the basically the same for std::memcpy and effective_memcpy when we refactor with const template, while in the original code, introducing effective_memcpy does improve the performance. 

Maybe the compiler will do some optimization on std::memcpy when it knows the `Length` in advance in compile time?

wenym1 replied:
My guess is correct. 

Just try compile the code on `compiler explorer`. If we know the length of `memcpy` in advance, the compiler will not call `memcpy` and instead generate several load instruction. If the length is provided at runtime passed from the function parameter, the generated assembly code will call `memcpy`.

https://godbolt.org/z/5z1z5WPY7

wenym1 commented on include/runtime/instance/memory.h:
```diff
@@ -241,16 +270,11 @@ class MemoryInstance {
   ///
   /// \returns void when success, ErrCode when failed.
-  template <typename T>
+  template <typename T, uint32_t const Length = sizeof(T)>
```
I have removed the `const`

wenym1 commented:
I have updated the code according to some comments. 

For the `memcpy`, I added the alignment check, and in case of any of `dest` and `src` is not aligned, it will use `memcpy` instead. 

Are there any new comments? @q82419 @ibmibmibm

ibmibmibm commented:
Could you just check that performance is improved by using `effective_memcpy`, instead of `std::memcpy`?
I doubt that matter.

wenym1 commented:
> Could you just check that performance is improved by using effective_memcpy, instead of std::memcpy?
I doubt that matter.

I have put a benchmark result in the discussion above. It seems that when we know the `Length` in advance, the compiler will do the optimization for us automatically, and therefore refactoring with const template is enough and we don't need the extra `effective_memcpy`. I have removed the new function and use memcpy back.

For the optimization done by compiler you can refer to https://godbolt.org/z/5z1z5WPY7

@ibmibmibm

ibmibmibm commented:
You need to squash your commit into one, and rebase it to master.

wenym1 commented:
@q82419 Could you revert the change request? It's blocking me from merging this PR.
